sharktank/sharktank/layers/paged_attention.py

```diff
@@ -367,64 +227,622 @@ def write_timestep(
+                transformer_block_count=transformer_block_count,
+                attn_head_count=sharded_attn_head_count,
+                attn_head_dim=attn_head_dim,
+                cache_partition_count=cache_partition_count,
+                block_seq_stride=block_seq_stride,
+                cache_dtype=cache_dtype,
+                device=device,
+            )
+            caches.append(cache)
+
+        self.caches = caches
+        self.cache_partition_count = cache_partition_count
+        self.devices = devices
+        self.attn_head_count = attn_head_count
+        self.attn_head_dim = attn_head_dim
+        self.block_seq_stride = block_seq_stride
+        self.transformer_block_count = transformer_block_count
+        self.shard_count = shard_count
+
+        self.unsharded_page_dims = [
+            self.transformer_block_count,
+            self.cache_partition_count,
+            self.block_seq_stride,
+            self.attn_head_count,
+            self.attn_head_dim,
+        ]
+
+        if self.devices is None:
+            self.devices = list(range(shard_count))
+
+    @property
+    def state_count(self):
+        return 1
+
+    def allocate(
+        self, page_count: int, devices: None = None
+    ) -> List[SplitPrimitiveTensor]:
+        assert devices is None
+        shards = [cache.allocate(page_count)[0] for cache in self.caches]
+        return [SplitPrimitiveTensor(ts=shards, shard_dim=1, devices=self.devices)]
+
+    def shard_state(self, state: List[torch.Tensor]) -> List[SplitPrimitiveTensor]:
+        assert len(state) == 1
+        page_table = state[0].unflatten(1, self.unsharded_page_dims)
+
+        shards = []
+        head_start = 0
+        for cache in self.caches:
+            head_end = head_start + cache.attn_head_count
+            shard = page_table[:, :, :, :, head_start:head_end]
+            shard = shard.flatten(1)
+            shards.append(shard)
+            head_start = head_end
+
+        return [SplitPrimitiveTensor(ts=shards, shard_dim=1, devices=self.devices)]
+
+    def unshard_state(self, state: List[SplitPrimitiveTensor]) -> List[torch.Tensor]:
+        assert len(state) == 1
+        assert state[0].shard_count == len(self.caches)
+
+        state = [
+            cache.unshard_state([shard])[0]
+            for cache, shard in zip(self.caches, state[0].shards)
+        ]
+        state = SplitPrimitiveTensor(ts=state, shard_dim=4, devices=self.devices)
+
+        return [ops.unshard(state)]
+
+    def unflatten_page_table(
+        self, state: List[SplitPrimitiveTensor]
+    ) -> List[SplitPrimitiveTensor]:
+        assert False
+
+    def read(
+        self,
+        state: List[SplitPrimitiveTensor],
+        *,
+        transformer_block_index: int,
+        page_ids: ReplicatedTensor,
+    ):
+        assert len(state) == 1
+        assert state[0].shard_count == self.shard_count
+
+        for device in state[0].devices:
+            assert device in page_ids.devices
```

Comment:
Do we not also want to make sure they're in the correct order?

Reply:
`ReplicatedTensor` has no behavioral different by device ordering as each tensor should contain identical data. We just need to grab the tensor placed on the correct device.

Reply:
Am I missing something? It doesn't look like the downstream code performs that grabbing. It zips them in positional order, rather than based on matching device.

Reply:
Updated for missing indirection